source/MaterialXGenMdl/Nodes/SourceCodeNodeMdl.cpp

```diff
@@ -0,0 +1,192 @@
+//
+// TM & (c) 2017 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#include <MaterialXGenMdl/Nodes/SourceCodeNodeMdl.h>
+#include <MaterialXGenShader/GenContext.h>
+#include <MaterialXGenShader/ShaderNode.h>
+#include <MaterialXGenShader/ShaderStage.h>
+#include <MaterialXGenShader/ShaderGenerator.h>
+#include <MaterialXGenShader/Util.h>
+
+namespace MaterialX
+{
+
+ShaderNodeImplPtr SourceCodeNodeMdl::create()
+{
+    return std::make_shared<SourceCodeNodeMdl>();
+}
+
+void SourceCodeNodeMdl::initialize(const InterfaceElement& element, GenContext& context)
+{
+    ShaderNodeImpl::initialize(element, context);
+
+    if (!element.isA<Implementation>())
+    {
+        throw ExceptionShaderGenError("Element '" + element.getName() + "' is not an Implementation element");
+    }
+
+    const Implementation& impl = static_cast<const Implementation&>(element);
+
+    // Get source code from either an attribute or a file.
+    _functionSource = impl.getAttribute("sourcecode");
+    if (_functionSource.empty())
+    {
+        const FilePath file(impl.getAttribute("file"));
+        if (!readFile(context.resolveSourceFile(file), _functionSource))
+        {
+            throw ExceptionShaderGenError("Can't find source file '" + file.asString() +
+                "' used by implementation '" + impl.getName() + "'");
+        }
+    }
+
+    // Find the function name to use
+    // If no function is given the source will be inlined.
+    _functionName = impl.getAttribute("function");
+    _inlined = _functionName.empty();
+    if (!_inlined)
+    {
+        // Make sure the function name is valid.
+        string validFunctionName = _functionName;
+        context.getShaderGenerator().getSyntax().makeValidName(validFunctionName);
+        if (_functionName != validFunctionName)
+        {
+            throw ExceptionShaderGenError("Function name '" + _functionName +
+                "' used by implementation '" + impl.getName() + "' is not a valid identifier.");
+        }
+    }
+    else
+    {
+        _functionSource.erase(std::remove(_functionSource.begin(), _functionSource.end(), '\n'), _functionSource.end());
+    }
+
+    // Set hash using the function name.
+    // TODO: Could be improved to include the full function signature.
+    _hash = std::hash<string>{}(_functionName);
+}
+
+void SourceCodeNodeMdl::emitFunctionDefinition(const ShaderNode&, GenContext& /*context*/, ShaderStage& /*stage*/) const
+{
+#if 0
```

Comment:
@niklasharrysson , this fixes up the sourcecode constantly being included. I think we should make it add an import but it works for now with the imports explicitly added.

Reply:
I think the reason why source code for function definitions is being emitted is that the library files with <implementation> element has not been fully updated yet. I have only updated implementations on the fly as I start testing them and the majority of implementations has not be updated yet, so they still reference OSL code. All of them should start using the "sourcecode" attribute and without the "function" attribute, which should resolve this as they are then always inlined.

We should probably add this back, so that we can still emit function definitions when really needed.